lib/graphql/lang/stack.ex

```diff
@@ -0,0 +1,32 @@
+defmodule GraphQL.Lang.Stack do
+
+  alias GraphQL.Lang.Stack 
+
+  # Path is leaf to root. This means we can append to the head of the list as
+  # we go deeper and pop from the head of the list as we go shallower. This
+  # uses Erlang's singly linked lists in an efficient way.
+  defstruct path: []
+
+  def parent(stack) do
+    if length(stack.path) >= 2 do
+      [_|parent] = stack
+      parent
+    else
+      nil
+    end
+  end 
+
+  def push(stack, node) do
+    %Stack{ stack | path: [node] ++ stack.path }
+  end
+
+  def pop(stack) do
```

Comment:
I can actually get rid of `pop`. The nice thing about an immutable stack is that you never need to pop. The function just returns and the stack in the scope _is_ the popped version.